### ISSUE TITLE
Crash on input after ctrl-z

### ISSUE THREAD

**Issue by redmanmale** [uTox/uTox#1072]
**Steps to reproduce**
* type some text into any edit
* cancel some text input via CTRL-Z
* type again

If that's not working try to type and cancel couple times.

**Env**
Win 7, 8.1, uTox v.0.16.1 and latest dev

**Comment by fcore117** [comment]
[utox_p_debug_log.txt](https://github.com/uTox/uTox/files/1346411/utox_p_debug_log.txt)

**Comment by redmanmale** [comment]
Crashed [here](https://github.com/uTox/uTox/blob/develop/src/ui/edit.c#L372) but it seems that memory is already corrupted there.

**Comment by g4jc** [comment]
ASAN:
```
==7839==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602002df47b8 at pc 0x555555658d93 bp 0x7fffffffd060 sp 0x7fffffffd050
READ of size 8 at 0x602002df47b8 thread T0
    #0 0x555555658d92 in edit_do (/usr/bin/utox+0x104d92)
    #1 0x555555663713 in edit_char (/usr/bin/utox+0x10f713)
    #2 0x5555556b042a  (/usr/bin/utox+0x15c42a)
    #3 0x5555555d59e3 in main (/usr/bin/utox+0x819e3)
    #4 0x7ffff707c8e9 in __libc_start_main (/lib64/libc.so.6+0x218e9)
    #5 0x5555555d62e9 in _start (/usr/bin/utox+0x822e9)

0x602002df47b8 is located 0 bytes to the right of 8-byte region [0x602002df47b0,0x602002df47b8)
allocated by thread T0 here:
    #0 0x7ffff72b8580 in __interceptor_realloc /var/tmp/portage/sys-devel/gcc-7.3.0-r6/work/gcc-7.3.0/libsanitizer/asan/asan_malloc_linux.cc:75
    #1 0x555555658ace in edit_do (/usr/bin/utox+0x104ace)
    #2 0x555555663713 in edit_char (/usr/bin/utox+0x10f713)
    #3 0x5555556b042a  (/usr/bin/utox+0x15c42a)
    #4 0x5555555d59e3 in main (/usr/bin/utox+0x819e3)
    #5 0x7ffff707c8e9 in __libc_start_main (/lib64/libc.so.6+0x218e9)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/usr/bin/utox+0x104d92) in edit_do
Shadow bytes around the buggy address:
  0x0c04805b68a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c04805b68b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c04805b68c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c04805b68d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c04805b68e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c04805b68f0: fa fa 07 fa fa fa 00[fa]fa fa fd fd fa fa fd fd
  0x0c04805b6900: fa fa fd fd fa fa fd fd fa fa 00 07 fa fa fd fd
  0x0c04805b6910: fa fa fd fd fa fa fd fd fa fa fd fd fa fa fd fd
  0x0c04805b6920: fa fa fd fd fa fa fd fd fa fa fd fd fa fa fd fd
  0x0c04805b6930: fa fa fd fd fa fa fd fd fa fa fd fd fa fa fd fd
  0x0c04805b6940: fa fa fd fd fa fa fd fd fa fa fd fd fa fa fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==7839==ABORTING
[Thread 0x7fffddabc700 (LWP 7992) exited]
[Thread 0x7fffe25e5700 (LWP 7987) exited]
[Thread 0x7fffe2e11700 (LWP 7986) exited]
[Thread 0x7fffe3086700 (LWP 7847) exited]
[Thread 0x7fffe319d700 (LWP 7846) exited]
[Thread 0x7fffe32ac700 (LWP 7845) exited]
[Thread 0x7ffff5d7aa40 (LWP 7839) exited]
--Type <RET> for more, q to quit, c to continue without paging--
```

Backtrace:
```
[Thread 0x7fffe5df0700 (LWP 19914) exited]
XLib Event:   Type UTF8_STRING || Prop TARGETS 
​
Thread 1 "utox" received signal SIGSEGV, Segmentation fault.
0x00007ffff7eab019 in free () from /lib64/libc.so.6
(gdb) bt full
#0  0x00007ffff7eab019 in free () at /lib64/libc.so.6
#1  0x00005555555bb4ba in edit_do (edit=0x5555556c1fe0 <edit_chat_msg_friend>, start=0, length=1, remove=false)
    at /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/ui/edit.c:379
        i = 4
        new = 0x5555559bac20
        history = 0x5555559baba0
#2  0x00005555555911eb in edit_char (ch=122, control=false, flags=16 '\020') at /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/ui/edit.c:764
        len = 1 '\001'
        p = 0x555555806800 <e_chat_msg_friend_data> "zaaaaaaaaa"
        edit = 0x5555556c1fe0 <edit_chat_msg_friend>
#3  0x00005555555aa597 in doevent (event=0x7fffffffdda0) at /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/xlib/event.c:551
        key = 122
        ev = 0x7fffffffdda0
        sym = 122
        buffer = L"z\000\xf6bf195b翿\000\000\x5c046300\x764ea5b5\x55967620啕\x55849f10啕\x5584b190啕\000"
        len = 1
#4  0x00005555555b2a43 in main (argc=1, argv=0x7fffffffdf58) at /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/xlib/main.c:855
        event = 
            {type = 2, xany = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193}, xkey = {type = 2, serial = 40271, send_event = 0, di--Type <RET> for more, q to quit, c to continue without paging--
splay = 0x555555849f10, window = 44040193, root = 1378, subwindow = 0, time = 37949772, x = 442, y = 523, x_root = 447, y_root = 973, state = 16, keycode = 52, same_screen = 1}, xbutton = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, root = 1378, subwindow = 0, time = 37949772, x = 442, y = 523, x_root = 447, y_root = 973, state = 16, button = 52, same_screen = 1}, xmotion = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, root = 1378, subwindow = 0, time = 37949772, x = 442, y = 523, x_root = 447, y_root = 973, state = 16, is_hint = 52 '4', same_screen = 1}, xcrossing = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, root = 1378, subwindow = 0, time = 37949772, x = 442, y = 523, x_root = 447, y_root = 973, mode = 16, detail = 52, same_screen = 1, focus = 1, state = 20}, xfocus = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, mode = 1378, detail = 0}, xexpose = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, x = 1378, y = 0, width = 0, height = 0, count = 37949772}, xgraphicsexpose = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, drawable = 44040193, x = 1378, y = 0, width = 0, height = 0, count = 37949772, major_code = 0, minor_code = 442}, xnoexpose = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, drawable = 44040193, major_code = 1378, minor_code = 0}, xvisibility = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, state = 1378}, xcreatewindow = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, parent = 44040193, window = 1378, x = 0, y = 0, width = 37949772, height = 0, border_width = 442, override_redirect = 523}, xdestroywindow = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378}, xunmap = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378, from_configure = 0}, xmap = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378, override_redirect = 0}, xmaprequest = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, parent = 44040193, window = 1378}, xreparent = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378, parent = 0, x = 37949772, y = 0, override_redirect = 442}, xconfigure = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378, x = 0, y = 0, width = 37949772, height = 0, border_width = 442, above = 4179003179455, override_redirect = 16}, xgravity = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378, x = 0, y = 0}, xresizerequest = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, width = 1378, height = 0}, xconfigurerequest = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, parent = 44040193, window = 1378, x = 0, y = 0, width = 37949772, height = 0, border_width = 442, above = 4179003179455, detail = 16, value_mask = 42--Type <RET> for more, q to quit, c to continue without paging--
94967297}, xcirculate = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, event = 44040193, window = 1378, place = 0}, xcirculaterequest = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, parent = 44040193, window = 1378, place = 0}, xproperty = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, atom = 1378, time = 0, state = 37949772}, xselectionclear = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, selection = 1378, time = 0}, xselectionrequest = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, owner = 44040193, requestor = 1378, selection = 0, target = 37949772, property = 2246267896250, time = 4179003179455}, xselection = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, requestor = 44040193, selection = 1378, target = 0, property = 37949772, time = 2246267896250}, xcolormap = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, colormap = 1378, new = 0, state = 0}, xclient = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, message_type = 1378, format = 0, data = {b = "L\021C\002\000\000\000\000\272\001\000\000\v\002\000\000\277\001\000", s = {4428, 579, 0, 0, 442, 0, 523, 0, 447, 0}, l = {37949772, 2246267896250, 4179003179455, 223338299408, 4294967297}}}, xmapping = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, request = 1378, first_keycode = 0, count = 0}, xerror = {type = 2, display = 0x9d4f, resourceid = 0, serial = 93824995335952, error_code = 1 '\001', request_code = 0 '\000', minor_code = 160 '\240'}, xkeymap = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, window = 44040193, key_vector = "b\005", '\000' <repeats 14 times>, "L\021C\002\000\000\000\000\272\001\000\000\v\002\000"}, xgeneric = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, extension = 44040193, evtype = 0}, xcookie = {type = 2, serial = 40271, send_event = 0, display = 0x555555849f10, extension = 44040193, evtype = 0, cookie = 1378, data = 0x0}, pad = {140733193388034, 40271, 0, 93824995335952, 44040193, 1378, 0, 37949772, 2246267896250, 4179003179455, 223338299408, 4294967297, 20, 0, 0, 0, 1630273536, 0, 11361, 0, 0, -1801568493808648192, 27, 39406385523326976}}
        should_launch_at_startup = 0 '\000'
        set_show_window = 0 '\000'
        allow_root = false
        xim = 0x5555558a2070
        action = 
--Type <RET> for more, q to quit, c to continue without paging--
          {__sigaction_handler = {sa_handler = 0x5555555b2278 <signal_handler>, sa_sigaction = 0x5555555b2278 <signal_handler>}, sa_mask = {__val = {0, 0, 140737353128817, 656, 140737353705952, 0, 140737352746105, 6, 140737352731083, 0, 0, 93824995262464, 4095, 18446744073709547520, 0, 0}}, sa_flags = 0, sa_restorer = 0x0}
        dndversion = 3
        gcval = 
          {function = 6, plane_mask = 16777215, foreground = 16777215, background = 0, line_width = 0, line_style = 0, cap_style = 0, join_style = 65280, fill_style = 1869833334, fill_rule = 1952802655, arc_mode = 1701669236, tile = 0, stipple = 0, ts_x_origin = 0, ts_y_origin = 0, font = 0, subwindow_mode = 1, graphics_exposures = 0, clip_x_origin = 0, clip_y_origin = 0, clip_mask = 140737336269568, dash_offset = -155394048, dashes = -1 '\377'}
        attr = 
          {x = 0, y = 0, width = 1920, height = 1080, border_width = 0, depth = 24, visual = 0x555555858890, root = 1378, class = 1, bit_gravity = 0, win_gravity = 1, backing_store = 0, backing_planes = 4294967295, backing_pixel = 0, save_under = 0, colormap = 32, map_installed = 0, map_state = 2, all_event_masks = 16416831, your_event_mask = 0, do_not_propagate_mask = 0, override_redirect = 0, screen = 0x55555584b810}
        xrcolor = {red = 0, green = 0, blue = 0, alpha = 0}
        root_return = 42
        child_return = 1224
        x_return = 32767
        y_return = -134395424
        width_return = 32767
        height_return = 4158828456
        i = 32767
```
Crash in while loop: https://github.com/uTox/uTox/blob/develop/src/ui/edit.c#L379

Apparently an off-by-one, but I've been unable to fix it while poking around.

**Comment by avoidr** [comment]
Closed by #1523.
